SR-2208: Improve failure diagnostics for apply expressions

xedin opened the pull request:
<!-- What's in this pull request? -->

Avoid mutating AST when trying to type-check OverloadSetRefExpr with assigned
contextual type because that degrades quality of diagnostics. If contextual type is
assigned to a binary expression and there is no distinct problem with arguments, 
patch tries to double check if there is an "exact match" candidate present in the
original set of callee candidates to distinguish between contextual result conversion
and everything else.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

Resolves #44815.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Diagnostics] SR-2208: Improve failure diagnostics for apply expressions

xedin commented:
/cc @rudkx

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5241,14 +5243,31 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
+          &listener);
+
+      if (type.hasValue())
+        fnType = type.getValue()->getRValueType();
```
With your change we'll now continue on when type.hasValue() is false, whereas I *think* before in these cases typeCheckChildIndependently() would have returned nullptr and we would have exited.

Is this really the right thing to do here?

xedin replied:
Ah, that's right, sorry! I'll make it so if there is no value, the nullptr is assigned, because it would be the same exact behavior as before.

rudkx replied:
I see. Thanks for the explanation.

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5241,14 +5243,31 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
-    if (!fnExpr) return true;
+
+    if (auto OSR = dyn_cast<OverloadSetRefExpr>(fnExpr)) {
+      ConcreteDeclRef decl = OSR->getReferencedDecl();
```
Are there ever cases where decl is non-nullptr here?

rudkx replied:
One approach would be to assign nullptr explicitly to decl after asserting the OSR->getReferencedDecl() is nullptr. I'd rename decl to referencedDecl as well.

It doesn't look like the ASTVerifier is actually doing any checks on these Exprs, so it would be nice to add a check there as well.

xedin replied:
Sounds good, I'll do that!

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5291,7 +5310,18 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
+    using Closeness = CalleeCandidateInfo::ClosenessResultTy;
+    calleeInfo.filterList([&](UncurriedCandidate candidate) -> Closeness {
+      auto resultType = candidate.getResultType();
+      if (resultType && !resultType->isEqual(fn->getResult()))
```
This seems like it's always going to filter out generic candidates since the result type will not be equal.

I'm wondering if there is an easy way to improve that. Offhand I don't know.

xedin replied:
I thought somewhere that there is TC.isConvertableTo method maybe that might help?

rudkx commented on test/Constraints/diagnostics.swift:
```diff
@@ -720,7 +719,8 @@ secondArgumentNotLabeled(10, 20)
 func testImplConversion(a : Int?) -> Bool {
   let someInt = 42
-  let a : Int = testImplConversion(someInt) // expected-error {{'testImplConversion' produces 'Bool', not the expected contextual result type 'Int'}}
+  let a : Int = testImplConversion(someInt) // expected-error {{argument labels '(_:)' do not match any available overloads}}
```
If you add an argument label here, what error do we produce?

xedin replied:
```
<REPL Input>:3:15: error: cannot invoke 'testImplConversion' with an argument list of type '(a: Int)'
```

If I do `let someInt: Int? = 42`, this is what it produces:
```
<REPL Input>:3:15: error: cannot convert value of type 'Bool' to specified type 'Int'
```

rudkx commented on validation-test/stdlib/FixedPointDiagnostics.swift.gyb:
```diff
@@ -2,13 +2,13 @@
 
 func testUnaryMinusInUnsigned() {
-  var a: UInt8 = -(1) // expected-error {{}} expected-note * {{}}
+  var a: UInt8 = -(1) // expected-error {{}} expected-note * {{}} expected-warning * {{}}
```
So these now produce an error and a warning?

What's the warning?

xedin replied:
Warning is that "a" is not used and can be replaced with "_".

rudkx replied:
The warning is new with your change? That seems unexpected.

xedin replied:
Yes, that warning is new, I will take a look why that's happening, it's emitted after "error".

xedin replied:
This actually happens only when function body is being diagnosed...

rudkx replied:
I'm not sure I follow. Isn't that warning emitted in a MiscDiagnostics.cpp pass? Are we now getting to that point after emitting an error whereas we didn't used to?

xedin replied:
Yes, somehow it's not respecting true that has been returned from diagnoseContextualConversionError, still digging through this...

xedin replied:
Just figured out what is going on with this warning, and here it is - since this is being diagnosed as contextual conversion error, actual pattern_binding declaration for "a" is going to be type-checked as "Int" because `typeCheckChildIndependently` applies solutions to AST...

Before changes (because "+" operation itself has been diagnosed to the type is not set):
```
(brace_stmt
  (pattern_binding_decl
    (pattern_typed type='UInt8'
      (pattern_named type='UInt8' 'a')
      (type_ident
        (component id='UInt8' bind=Swift.(file).UInt8)))
    (prefix_unary_expr type='<<error type>>' location=<REPL Input>:2:19 range=[<REPL Input>:2:19 - line:2:22]
      (overloaded_decl_ref_expr type='<<error type>>' location=<REPL Input>:2:19 range=[<REPL Input>:2:19 - line:2:19] name=- #decls=7 specialized=no function_ref=unapplied
      Swift.(file).-
      Swift.(file).-
      Swift.(file).-
      Swift.(file).-
      Swift.(file).-
      Swift.(file).-
      Swift.(file).SignedNumber.-)
      (paren_expr type='<<error type>>' location=<REPL Input>:2:21 range=[<REPL Input>:2:20 - line:2:22]
        (integer_literal_expr type='<<error type>>' location=<REPL Input>:2:21 range=[<REPL Input>:2:21 - line:2:21] value=1))))

  (var_decl "a" type='UInt8' access=private storage_kind=stored)
)
```
 After changes (because "+" is considered valid and contextual conversion is invalid):

```
(brace_stmt
  (pattern_binding_decl
    (pattern_typed type='UInt8'
      (pattern_named type='UInt8' 'a')
      (type_ident
        (component id='UInt8' bind=Swift.(file).UInt8)))
    (prefix_unary_expr type='Int' location=<REPL Input>:2:19 range=[<REPL Input>:2:19 - line:2:22]
      (declref_expr type='(Int) -> Int' location=<REPL Input>:2:19 range=[<REPL Input>:2:19 - line:2:19] decl=Swift.(file).- [with Int[Int: SignedNumber module Swift], Int] function_ref=unapplied specialized=no)
      (paren_expr type='(Int)' location=<REPL Input>:2:21 range=[<REPL Input>:2:20 - line:2:22]
        (call_expr implicit type='Int' location=<REPL Input>:2:21 range=[<REPL Input>:2:21 - line:2:21]  arg_labels=_builtinIntegerLiteral:
          (constructor_ref_call_expr implicit type='(Int2048) -> Int' location=<REPL Input>:2:21 range=[<REPL Input>:2:21 - line:2:21]
            (declref_expr implicit type='(Int.Type) -> (Int2048) -> Int' location=<REPL Input>:2:21 range=[<REPL Input>:2:21 - line:2:21] decl=Swift.(file).Int.init(_builtinIntegerLiteral:) function_ref=single specialized=no)
            (type_expr implicit type='Int.Type' location=<REPL Input>:2:21 range=[<REPL Input>:2:21 - line:2:21] typerepr='Int'))
          (tuple_expr implicit type='(_builtinIntegerLiteral: Int2048)' location=<REPL Input>:2:21 range=[<REPL Input>:2:21 - line:2:21] names=_builtinIntegerLiteral
            (integer_literal_expr type='Int2048' location=<REPL Input>:2:21 range=[<REPL Input>:2:21 - line:2:21] value=1))))))

  (var_decl "a" type='UInt8' access=private storage_kind=stored)
)
```

This looks to be caused by the fact that `diagnoseContextualConversionError` doesn't erase type of the expression after it being diagnosed as erroneous...

Do you think I should try and address it in that patch?

xedin replied:
Even with current master If I do this
```
(swift) func foo2() {  
   var b: UInt32 = UInt8(1)                                                                                                                                                                                                                                                                
}
```
I get back:

```
<REPL Input>:2:17: error: cannot convert value of type 'UInt8' to specified type 'UInt32'
 var b:UInt32 = UInt8(1)}
                ^~~~~~~~
                UInt32( )
<REPL Input>:2:6: warning: initialization of variable 'b' was never used; consider replacing with assignment to '_' or removing it
 var b:UInt32 = UInt8(1)}
 ~~~~^
 _
```

rudkx reviewed (commented):
Looks pretty good. A few questions/comments.

xedin commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5241,14 +5243,31 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
       (isUnresolvedOrTypeVarType(fnExpr->getType()) ||
        (fnExpr->getType()->is<AnyFunctionType>() &&
         fnExpr->getType()->hasUnresolvedType()))) {
+    // FIXME: Prevent typeCheckChildIndependently from transforming expressions,
```
Yes, that's exactly right, the same situation as with InOutExpr from https://github.com/apple/swift/pull/5105 , once typeCheckChildIndependency is improved, we can remove this workaround.

rudkx replied:
Okay.

rudkx commented:
@swift-ci Please smoke test

Run a test pass to verify the current PR.

rudkx commented:
@swift-ci Please smoke test

xedin commented:
@rudkx I've pushed changes for getReferencedDecl(), changed isEqual to use CS->matchTypes instead which should handle generic parameters properly, and rebased everything with the latest master.

xedin commented:
@rudkx I've force pushed changes related to generic types.

rudkx commented:
@swift-ci Please smoke test

rudkx commented on test/expr/closure/closures.swift:
```diff
@@ -33,7 +33,7 @@ func funcdecl5(_ a: Int, _ y: Int) {
   func6({$0 + $1})       // Closure with two named anonymous arguments
   func6({($0) + $1})    // Closure with sequence expr inferred type
-  func6({($0) + $0})    // // expected-error {{binary operator '+' cannot be applied to two '(Int, Int)' operands}} expected-note {{expected an argument list of type '(Int, Int)'}}
+  func6({($0) + $0})    // // expected-error {{binary operator '+' cannot be applied to two '(Int, Int)' operands}} expected-note {{overloads for '+' exist with these partially matching parameter lists}}
```
Hmm, this is unfortunate. I suspect it's due to allowing the generic overloads in as well?

rudkx commented:
The new diagnostics here are much better than before, despite not being perfect. Let's take this and we can iterate on further improvements in the future (perhaps with more infrastructure in place to make it easier to improve these).

Thanks!
